cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRegisterer.java

```diff
@@ -156,7 +161,101 @@ public Pipeline registerPlugins(ETLConfig config, Class errorDatasetType, Datase
+      // check if this connections's from and to belong to actual stage
+      Preconditions.checkState(
+        stageNamesFromConfig.contains(connection.getFrom()),
+        String.format("Connection from : %s does not belong to an actual stage name, please check the config",
+                      connection.getFrom()));
+      Preconditions.checkState(
+        stageNamesFromConfig.contains(connection.getTo()),
+        String.format("Connection to : %s does not belong to an actual stage name, please check the config",
+                      connection.getFrom()));
+      Preconditions.checkState(stageNamesFromConfig.contains(connection.getTo()));
+      if (stageConnections.containsKey(connection.getFrom())) {
+        stageConnections.get(connection.getFrom()).add(connection.getTo());
+      } else {
+        List<String> destinations = new ArrayList<>();
+        destinations.add(connection.getTo());
+        stageConnections.put(connection.getFrom(), destinations);
+      }
+    }
+
+    Set<String> visited = new HashSet<>();
+    visited.add(config.getSource().getName());
+    Set<String> sinksFromConfig = new HashSet<>();
+    for (ETLStage sink : config.getSinks()) {
+      sinksFromConfig.add(sink.getName());
+    }
+    connectionsReachabilityValidation(stageConnections, config,
+                                      config.getSource().getName(), visited, new HashSet<String>(), sinksFromConfig);
+    return stageConnections;
+  }
+
+  private static void connectionsReachabilityValidation(
+    Map<String, List<String>> mapStageToConnections, ETLConfig config,
+    String stageName, Set<String> visited, Set<String> sinksVisited, Set<String> sinksFromConfig) {
+
+    if (mapStageToConnections.get(stageName) == null) {
+      // check if this stage is a sink, if its not a sink, throw an exception.
+      if (!sinksFromConfig.contains(stageName)) {
+        throw new IllegalArgumentException(
+          String.format(
+            "Stage : %s is not connected to any transform or sink, please check the config", stageName));
+      }
+      // this is a sink, add it to sinks visited set
+      sinksVisited.add(stageName);
+      return;
+    }
+
+    for (String nextConnection : mapStageToConnections.get(stageName)) {
+      if (visited.contains(nextConnection)) {
+        // already been visited, cycle exists, throw exception
+        throw new IllegalArgumentException(
+          String.format(
+            "Connection %s --> %s causes a cycle, Graph has to be a DAG, " +
+              "please check the graph connections", stageName, nextConnection));
+      }
+      HashSet<String> nextVisited = new HashSet<>(visited);
+      nextVisited.add(nextConnection);
+      connectionsReachabilityValidation(mapStageToConnections, config, nextConnection, nextVisited,
+                                        sinksVisited, sinksFromConfig);
+    }
+    // check if we have visited all the sinks, check this only when we are at source iteration level
```

Comment:
like the comment says, this only needs to be checked after the dag traversal is complete.  I think we should be able to move it out of this method so that it is not checked with each recursive call.  The original caller has access to sinksVisited so it should be doable there.